### PR TITLE
Use core/app-starters-security-common for security management

### LINKED ISSUE #24: Use core/app-starters-security-common for security management.
Leverage the [centralized mechanism]( https://github.com/spring-cloud-stream-app-starters/app-starters-release/issues/161) for security control introduced.

To be applied only for Boot 2.1+ applications!

Related to:  https://github.com/spring-cloud-stream-app-starters/app-starters-release/issues/161
Related to: https://github.com/spring-cloud/spring-cloud-app-starters-maven-plugins/issues/36

### PR DESCRIPTION
**Pull request by tzolov**
- Add core/app-starters-security-common dependency.
 - Remove the internal security management in favor of using the core/app-starters-security-common module.
 - Replace local enableSecurity and enableCsrf properties by and spring.cloud.stream.security.enabled and spring.cloud.stream.security.csrf-enabled from SecurityCommonAutoConfigurationProperties (inside core/app-starters-security-common).
 - Adjust existing tests with the new properties.

Resolves #24

Part of https://github.com/spring-cloud-stream-app-starters/app-starters-release/issues/161
Related to https://github.com/spring-cloud/spring-cloud-app-starters-maven-plugins/issues/36

### CODE REVIEW

File: http-app-dependencies/pom.xml
Comment (artembilan) [inline http-app-dependencies/pom.xml]: Why it is not in the `test` scope?
As far as we know this dependency is going to appear in all the generated apps. So, no need to specify it explicitly here.

If you would like to be sure in tests in this project before generation, then it must go to `test` scope.

File: spring-cloud-starter-stream-source-http/README.adoc
Comment (artembilan) [inline spring-cloud-starter-stream-source-http/README.adoc]: Why do you include these properties here explicitly?
I don't think that it is a responsibility of this HTTP app to expose them.
Or are you going to say the same in all the other apps starters?

Thanks

File: spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java
Comment (artembilan) [inline spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java]: And why do you include this `SecurityCommonAutoConfigurationProperties` here?
Isn't it included automatically by the `SecurityCommonAutoConfiguration`, so then it must not be here on the HTTP Source level.

File: spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties
Comment (artembilan) [inline spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties]: OK. I agree with this approach and I believe that's how configuration properties are appeared in the README.

Only the problem that we are going to do this kind of change in all the app starters.

Can't we just live without those properties and just mention in the docs that security is disabled by default and can be turned on using particular properties?

I mean that's not what I expected from this PR. 

We should remove all the security related code and if we wish we can include the common module for tests coverage.

Let me know if I am missing something.

Thanks

File: spring-cloud-starter-stream-source-http/src/test/java/org/springframework/cloud/stream/app/http/source/HttpSourceTests.java
Comment (artembilan) [inline spring-cloud-starter-stream-source-http/src/test/java/org/springframework/cloud/stream/app/http/source/HttpSourceTests.java]: I don't think we need to specify this property explicitly.
We definitely have to be sure that default behavior is with the disabled security.

So, this change has to be removed.

**Review by artembilan — CHANGES_REQUESTED** [review]
That's a review from me.

File: spring-cloud-starter-stream-source-http/README.adoc
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/README.adoc]: Currently the `app-starters-security-common for security` is applied only to the `http-source` app stater. 
Since i've removed the analogous http security properties it though it might be useful to list the new one,  although they come form an external module. 

What do you think?  I can remove it if you think that is inconsistent with other porops policies?

File: http-app-dependencies/pom.xml
Comment (tzolov) [reply http-app-dependencies/pom.xml]: > As far as we know this dependency is going to appear in all the generated apps. 

This is not true! Because we found that the Management security in 2.1 + enables info and health by default, i've removed it from all generated apps. 
That is way i have to add it as non-test dependency for the http-source app.

File: spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties]: As i mentioned above, at the moment only the http-source uses the security module autoconfiguration.

File: spring-cloud-starter-stream-source-http/src/test/java/org/springframework/cloud/stream/app/http/source/HttpSourceTests.java
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/src/test/java/org/springframework/cloud/stream/app/http/source/HttpSourceTests.java]: +1

File: spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/src/main/java/org/springframework/cloud/stream/app/http/source/HttpSourceConfiguration.java]: +1

File: spring-cloud-starter-stream-source-http/README.adoc
Comment (artembilan) [reply spring-cloud-starter-stream-source-http/README.adoc]: Yes, please, remove it. 
Since we have now a general solution which covers this HTTP Source as well, we don't need to duplicate a logic and docs here.
Everything is covered from that common module perspective.

As I said: we just need to have a proper docs in that module, plus on the App Starters level: https://docs.spring.io/spring-cloud-stream-app-starters/docs/Darwin.SR2/reference/htmlsingle/

File: spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties
Comment (artembilan) [reply spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties]: It must not. Since that module is auto-configuration anyway and it is going to appear in each generated app there is no reason to duplicate the logic.
What we need 3is just remove home-made logic in this project since everything is covered by that new module.

We can leave tests here for better coverage and then dependency must go to the `test` scope.

Not sure in your logic to duplicate that module on the `compile` level...

File: http-app-dependencies/pom.xml
Comment (tzolov) [reply http-app-dependencies/pom.xml]: Ok the final state is that the dependency will be present in all app starters, but disabling the security only for Info, Health, LogFile and Bindings endpoints by default.

Still `test` scope makes perfect sense now.

File: spring-cloud-starter-stream-source-http/README.adoc
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/README.adoc]: Ok done

File: spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/src/main/resources/META-INF/spring-configuration-metadata-whitelist.properties]: +1

File: spring-cloud-starter-stream-source-http/src/test/java/org/springframework/cloud/stream/app/http/source/HttpSourceTests.java
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/src/test/java/org/springframework/cloud/stream/app/http/source/HttpSourceTests.java]: Because after latest updates the `core\app-starters-security-common` disables security only for Info, Health, LogFile and Bindings by default the `http` app uses the `config\application.yml` to set `spring.cloud.stream.security.enabled` to `false`

**Comment by tzolov** [comment]
Depends on: https://github.com/spring-cloud-stream-app-starters/core/pull/59

File: http-app-dependencies/pom.xml
Comment (artembilan) [reply http-app-dependencies/pom.xml]: I still insist that we should not have these dependencies for testing in the BOM.

Just move the to the management in the root pom which is not going to be a part of target application when we import this BOM there.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-cloud-starter-stream-source-http/src/main/resources/config/application.yml
Comment (ilayaperumalg) [inline spring-cloud-starter-stream-source-http/src/main/resources/config/application.yml]: I think we want to change this to `stream-app-starters`, no?

**Review by artembilan — CHANGES_REQUESTED** [review]
Can we also move an `httpclient` dependency out of the BOM as well?
Otherwise LGTM, although we still need to agree with the property name for security...

File: spring-cloud-starter-stream-source-http/src/main/resources/config/application.yml
Comment (tzolov) [reply spring-cloud-starter-stream-source-http/src/main/resources/config/application.yml]: fixed

**Review by artembilan — APPROVED** [review]
LGTM.
Although we are still waiting for the final decision about the property name.

**Comment by dturanski** [comment]
I'm getting ```Failures:
[ERROR]   HttpSourceTests$CsrfEnabledTests.testText:257 expected:<401 UNAUTHORIZED> but was:<202 ACCEPTED>```

**Comment by artembilan** [comment]
> I'm getting Failures:

Let's wait for upstream PR being merged!

Too much confusion with Maven local and potentially missed builds...

**Comment by tzolov** [comment]
@dturanski do you still seeing the `HttpSourceTests$CsrfEnabledTests.testText:257` failures? 
`./mvnw clean install -U` passes successfully for me.

**Comment by dturanski** [comment]
Works for me now

**Pull request merged**
